beam/core/src/main/java/cz/o2/proxima/beam/direct/io/CommitLogReadFn.java

```diff
@@ -0,0 +1,168 @@
+/**
+ * Copyright 2017-2020 O2 Czech Republic, a.s.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.o2.proxima.beam.direct.io;
+
+import cz.o2.proxima.direct.commitlog.CommitLogReader;
+import cz.o2.proxima.direct.commitlog.LogObserver;
+import cz.o2.proxima.direct.commitlog.ObserveHandle;
+import cz.o2.proxima.direct.commitlog.Offset;
+import cz.o2.proxima.repository.Repository;
+import cz.o2.proxima.repository.RepositoryFactory;
+import cz.o2.proxima.storage.Partition;
+import cz.o2.proxima.storage.StreamElement;
+import cz.o2.proxima.storage.commitlog.Position;
+import cz.o2.proxima.time.Watermarks;
+import cz.o2.proxima.util.ExceptionUtils;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.SerializableCoder;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.Impulse;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.splittabledofn.RestrictionTracker;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+/** A {@link DoFn} that reads from a {@link CommitLogReader} using splittable DoFn. */
+@DoFn.UnboundedPerElement
+public class CommitLogReadFn extends DoFn<byte[], StreamElement> {
+
+  public static PTransform<PBegin, PCollection<StreamElement>> reading(
+      String observeName, long limit, Repository repo, CommitLogReader reader) {
+    return new PTransform<PBegin, PCollection<StreamElement>>() {
```

Comment:
can we have a static inner class instead? these kind of constructs are hard to debug from stack traces

maybe we should rather stick to beam's convetion (which is the exact opposite) -> top level class should be a PTransform and DoFn should be a static inner class

Reply:
I'm not sure how transposing the class composition helps readability of the stack traces, but I don't have any issues with creating a `CommitLogRead` PTransform.

Reply:
So :+1:

Reply:
Fixed